dotnet/src/Microsoft.Agents.AI.Abstractions.CosmosNoSql/CosmosChatMessageStore.cs

```diff
@@ -0,0 +1,853 @@
+    [UnconditionalSuppressMessage("AOT", "IL3050:Calling members annotated with 'RequiresDynamicCodeAttribute' may break functionality when AOT compiling", Justification = "ChatMessage deserialization is controlled")]
+    public override async Task<IEnumerable<ChatMessage>> GetMessagesAsync(CancellationToken cancellationToken = default)
+    {
+#pragma warning disable CA1513 // Use ObjectDisposedException.ThrowIf - not available on all target frameworks
+        if (this._disposed)
+        {
+            throw new ObjectDisposedException(GetType().FullName);
+        }
+#pragma warning restore CA1513
+
+        // Use type discriminator for efficient queries
+        var query = new QueryDefinition("SELECT * FROM c WHERE c.conversationId = @conversationId AND c.type = @type ORDER BY c.timestamp ASC")
+            .WithParameter("@conversationId", this._conversationId)
+            .WithParameter("@type", "ChatMessage");
+
+        var iterator = this._container.GetItemQueryIterator<CosmosMessageDocument>(query, requestOptions: new QueryRequestOptions
+        {
+            PartitionKey = this._partitionKey,
+            MaxItemCount = this.MaxItemCount // Configurable query performance
+        });
+
+        var messages = new List<ChatMessage>();
+
+        while (iterator.HasMoreResults)
+        {
+            var response = await iterator.ReadNextAsync(cancellationToken).ConfigureAwait(false);
+
+            foreach (var document in response)
+            {
+                if (!string.IsNullOrEmpty(document.Message))
+                {
+                    var message = JsonSerializer.Deserialize<ChatMessage>(document.Message, s_defaultJsonOptions);
+                    if (message != null)
+                    {
+                        messages.Add(message);
+                    }
+                }
+            }
+        }
+
+        return messages;
+    }
+
+    /// <inheritdoc />
+    [UnconditionalSuppressMessage("Trimming", "IL2026:Members annotated with 'RequiresUnreferencedCodeAttribute' require dynamic access", Justification = "ChatMessage serialization is controlled")]
+    [UnconditionalSuppressMessage("AOT", "IL3050:Calling members annotated with 'RequiresDynamicCodeAttribute' may break functionality when AOT compiling", Justification = "ChatMessage serialization is controlled")]
+    public override async Task AddMessagesAsync(IEnumerable<ChatMessage> messages, CancellationToken cancellationToken = default)
+    {
+        if (messages is null)
+        {
+            throw new ArgumentNullException(nameof(messages));
+        }
+
+#pragma warning disable CA1513 // Use ObjectDisposedException.ThrowIf - not available on all target frameworks
+        if (this._disposed)
+        {
+            throw new ObjectDisposedException(GetType().FullName);
+        }
+#pragma warning restore CA1513
+
+        var messageList = messages.ToList();
```

Comment:
In many cases messages will already be an IReadOnlyCollection<ChatMessage>, so casting to it avoids an enumeration of the IEnumerable.
```suggestion
        var messageList = messages as IReadOnlyCollection<ChatMessage> ?? messages.ToList();
```